src/r1cs/tests.rs

```diff
@@ -176,132 +160,43 @@ For K = 1:
 // y variables are constrained to be a valid shuffle of the x variables.
 struct KShuffleGadget {}
 
-/// Represents an error during the proof creation of verification for a KShuffle gadget.
-#[derive(Fail, Copy, Clone, Debug, Eq, PartialEq)]
-pub enum KShuffleError {
-    /// Error in the constraint system creation process
-    #[fail(display = "Invalid KShuffle constraint system construction")]
-    InvalidR1CSConstruction,
-    /// Occurs when there are insufficient generators for the proof.
-    #[fail(display = "Invalid generators size, too few generators for proof")]
-    InvalidGeneratorsLength,
-    /// Occurs when verification of an [`R1CSProof`](::r1cs::R1CSProof) fails.
-    #[fail(display = "R1CSProof did not verify correctly.")]
-    VerificationError,
-}
-
-impl From<R1CSError> for KShuffleError {
-    fn from(e: R1CSError) -> KShuffleError {
-        match e {
-            R1CSError::InvalidGeneratorsLength => KShuffleError::InvalidGeneratorsLength,
-            R1CSError::MissingAssignment => KShuffleError::InvalidR1CSConstruction,
-            R1CSError::VerificationError => KShuffleError::VerificationError,
-        }
-    }
-}
-
 impl KShuffleGadget {
-    pub fn fill_cs<CS: ConstraintSystem>(
-        cs: &mut CS,
-        x: Vec<(Variable, Assignment)>,
-        y: Vec<(Variable, Assignment)>,
-    ) -> Result<(), KShuffleError> {
+    pub fn fill_cs<CS: ConstraintSystem>(cs: &mut CS, x: Vec<Variable>, y: Vec<Variable>) {
```

Comment:
Another thought on the shuffle example: since the shuffle uses a challenge that's bound only to the HL witness variables, it's not really safely composable with other gadgets, so perhaps we should change the code to be a `KShuffleProof` with `prove`, `verify` methods and an internal `gadget` function? This way we would have an opportunity to demonstrate not just making a gadget function, but also how to encapsulate the ProverCS / VerifierCS setup, and to talk about what the challenges are bound to and why it matters.

Reply:
👍